presto-main/src/main/java/io/prestosql/operator/exchange/PageReference.java

```diff
@@ -44,11 +48,26 @@ public long getRetainedSizeInBytes()
 
     public Page removePage()
     {
-        int referenceCount = this.referenceCount.decrementAndGet();
+        int referenceCount = REFERENCE_COUNT_UPDATER.decrementAndGet(this);
         checkArgument(referenceCount >= 0, "Page reference count is negative");
         if (referenceCount == 0) {
-            onFree.run();
+            onPageReleased.accept(page.getRetainedSizeInBytes());
         }
         return page;
     }
+
+    @Override
+    public String toString()
+    {
+        return toStringHelper(this)
+                .add("size", getRetainedSizeInBytes())
+                .add("referenceCount", referenceCount)
+                .toString();
+    }
+
+    public static LongConsumer releaseListenerForExchangeMemoryManager(LocalExchangeMemoryManager memoryManager)
```

Comment:
this is binding `PageReference` code with `LocalExchangeMemoryManager` unnecessarily.

Reply:
This avoids accidentally creating a megamorphic call site inside of `removePage()`. Currently, there are only two implementations of used- this one and the anonymous one created by `PassthroughExchanger`. If each site created their own lambda it would be megamorphic.

Reply:
This avoids accidentally creating a megamorphic call site inside of `removePage()`. Currently, there are only two implementations used- this one and the anonymous one created by `PassthroughExchanger`. If each exchanger created their own lambda it would be megamorphic.

Reply:
This is certainly an interesting idea, but does this actually make a difference? Presto is pretty much built around pushing megamorphic call sites to page-sized granularities. If this needs to be specialized like this, we probably have much bigger issues throughout the entire engine.

Reply:
I agree with @erichwang . We should not obscure readability for negligible perf improvements